### PR TITLE
Refactored the connections map from HashMap to DashMap

### PR DESCRIPTION
**Pull request by barshaul**
This PR changes the connection map data structure to DashMap, which helps reduce write lock contention on the connection container. The write lock on the connection container now will only be used when both slots and connection maps are being replaced.

### CODE REVIEW

File: redis/Cargo.toml [hunk 1]
```diff
@@ -63,6 +63,9 @@ crc16 = { version = "0.4", optional = true }
 rand = { version = "0.8", optional = true }
 derivative = { version = "2.2.0", optional = true }
 
+# Only needed for async cluster
+dashmap = { version = "6.0.1", optional = true }
```
Comment (eifrah-aws) [inline redis/Cargo.toml:67]: why are we locking into a a specific version? why not `6.0` ?

File: redis/src/cluster_async/connections_container.rs [hunk 2]
```diff
@@ -86,11 +85,12 @@ pub(crate) enum ConnectionType {
     PreferManagement,
 }
 
-pub(crate) struct ConnectionsMap<Connection>(pub(crate) HashMap<ArcStr, ClusterNode<Connection>>);
+pub(crate) struct ConnectionsMap<Connection>(pub(crate) DashMap<ArcStr, ClusterNode<Connection>>);
```
Comment (eifrah-aws) [inline redis/src/cluster_async/connections_container.rs:88]: Do we still need `ArcStr`? IIRC, the entire interface of `DashMap` is non-mutable, so I am not sure its needed

File: redis/src/cluster_async/connections_container.rs [hunk 3]
```diff
@@ -302,13 +309,13 @@ mod tests {
             }
         }
     }
-    fn remove_nodes(container: &mut ConnectionsContainer<usize>, addresss: &[&str]) {
+    fn remove_nodes(container: &ConnectionsContainer<usize>, addresss: &[&str]) {
```
Comment (eifrah-aws) [inline redis/src/cluster_async/connections_container.rs:312]: typo should be: `addresses`

File: redis/Cargo.toml (hunk 1 above)
Comment (barshaul) [reply redis/Cargo.toml:67]: It doesn't matter, it still means 6.0.1 and above. but I can change it

Comment (barshaul) [reply redis/Cargo.toml:67]: see https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html

File: redis/src/cluster_async/connections_container.rs (hunk 2 above)
Comment (barshaul) [reply redis/src/cluster_async/connections_container.rs]: Changed

**Comment by barshaul** [comment]
@eifrah-aws ready

File: redis/Cargo.toml (hunk 1 above)
Comment (eifrah-aws) [reply redis/Cargo.toml]: I am ok with this, I wondered why the ".1" does it contain a fix or was it just the latest at the time picked the crate ...

**Review by eifrah-aws — APPROVED** [review]

**Pull request merged**